File: inference.py
import torch
from PIL import Image
import numpy as np
import argparse

categories = ["music", "noise", "speech"]

def load_model(model_path="audio_classifier.pt"):
    """Loads the TorchScript model."""
    model = torch.jit.load(model_path)
    model.eval()
    return model

def preprocess_image(image, start_x, width, height):
    """Extracts a window from the image and converts it into a tensor."""
    image_window = image.crop((start_x, 0, start_x + width, height))
    image_array = np.array(image_window).astype(np.float32) / 255.0  # Normalize
    image_array = image_array.transpose(2, 0, 1)  # Convert to [C, H, W]
    return torch.tensor(image_array).unsqueeze(0)  # Add batch dimension

def predict(model, file_path):
    image = Image.open(file_path).convert('RGB')
    width, height = image.size
    step_size = 14
    window_size = 28
    
    for start_x in range(0, width - window_size + 1, step_size):
        input_tensor = preprocess_image(image, start_x, window_size, height)
        with torch.no_grad():
            output = model(input_tensor)
            prediction = torch.argmax(output, dim=1).item()
        print(f"{start_x}: {categories[prediction]}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Run inference on an image file with sliding windows.")
    parser.add_argument("file_path", type=str, help="Path to the image file.")
    parser.add_argument("--model", type=str, default="audio_classifier.pt", help="Path to the model file.")
    
    args = parser.parse_args()
    model = load_model(args.model)
    predict(model, args.file_path)

File: inference.js
const ort = require('onnxruntime-node'); // or: onnxruntime-web
const { createCanvas, loadImage } = require('canvas');
const { Tensor } = ort;

const categories = ['music', 'noise', 'speech'];
const modelPath = 'audio_classifier.onnx';

async function loadModel() {
	return await ort.InferenceSession.create(modelPath);
}

async function preprocessImage(imagePath, startX, width, height) {
	const image = await loadImage(imagePath);
	const canvas = createCanvas(width, height);
	const ctx = canvas.getContext('2d');
	ctx.drawImage(image, startX, 0, width, height, 0, 0, width, height);

	const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height).data;
	const floatData = new Float32Array(3 * width * height);

	for (let i = 0; i < width * height; i++) {
		floatData[i] = imageData[i * 4] / 255.0; // Red channel
		floatData[i + width * height] = imageData[i * 4 + 1] / 255.0; // Green channel
		floatData[i + 2 * width * height] = imageData[i * 4 + 2] / 255.0; // Blue channel
	}

	return new Tensor('float32', floatData, [1, 3, height, width]);
}

async function predict(model, imagePath) {
	const image = await loadImage(imagePath);
	const windowSize = 28;
	const stepSize = windowSize / 2;
	const height = image.height;
	const width = image.width;

	for (let startX = 0; startX + windowSize <= width; startX += stepSize) {
		const inputTensor = await preprocessImage(
			imagePath,
			startX,
			windowSize,
			height
		);
		const results = await model.run({ input: inputTensor });
		const output = results.output.data;
		const predictionIndex = output.indexOf(Math.max(...output));
		console.log(`${startX}: ${categories[predictionIndex]}`);
	}
}

(async () => {
	const model = await loadModel();
	const imageFile = process.argv[2];
	if (!imageFile) {
		console.log('Please provide an image file path.');
		process.exit(1);
	}
	await predict(model, imageFile);
})();
